lib/handlers/authorize-handler.js

```diff
@@ -0,0 +1,310 @@
+
+          this.updateResponse(response, redirectUri, state);
+
+          throw e;
+        });
+    });
+};
+
+/**
+ * Generate authorization code.
+ */
+
+AuthorizeHandler.prototype.generateAuthorizationCode = function() {
+  if (this.model.generateAuthorizationCode) {
+    return Promise.try(this.model.generateAuthorizationCode);
+  }
+
+  return tokenUtil.generateRandomToken();
+};
+
+/**
+ * Get authorization code lifetime.
+ */
+
+AuthorizeHandler.prototype.getAuthorizationCodeLifetime = function() {
+  var expires = new Date();
+
+  expires.setSeconds(expires.getSeconds() + this.authorizationCodeLifetime);
+
+  return expires;
+};
+
+/**
+ * Get the client from the model.
+ */
+
+AuthorizeHandler.prototype.getClient = function(request) {
+  var clientId = request.body.client_id || request.query.client_id;
+
+  if (!clientId) {
+    throw new InvalidRequestError('Missing parameter: `client_id`');
+  }
+
+  if (!is.vschar(clientId)) {
+    throw new InvalidRequestError('Invalid parameter: `client_id`');
+  }
+
+  var redirectUri = request.body.redirect_uri || request.query.redirect_uri;
+
+  if (redirectUri && !is.uri(redirectUri)) {
+    throw new InvalidRequestError('Invalid request: `redirect_uri` is not a valid URI');
+  }
+
+  return Promise.try(this.model.getClient, clientId)
+    .then(function(client) {
+      if (!client) {
+        throw new InvalidClientError('Invalid client: client credentials are invalid');
+      }
+
+      if (!client.grants) {
+        throw new InvalidClientError('Invalid client: missing client `grants`');
+      }
+
+      if (!_.contains(client.grants, 'authorization_code')) {
+        throw new UnauthorizedClientError('Unauthorized client: `grant_type` is invalid');
+      }
+
+      if (!client.redirectUri) {
+        throw new InvalidClientError('Invalid client: missing client `redirectUri`');
+      }
+
+      if (redirectUri && client.redirectUri !== redirectUri) {
+        throw new InvalidClientError('Invalid client: `redirect_uri` does not match client value');
+      }
+
+      return client;
+    });
+};
+
+/**
+ * Get scope from the request.
+ */
+
+AuthorizeHandler.prototype.getScope = function(request) {
+  var scope = request.body.scope || request.query.scope;
+
+  if (!is.nqschar(scope)) {
+    throw new InvalidScopeError('Invalid parameter: `scope`');
+  }
+
+  return scope;
+};
+
+/**
+ * Get state from the request.
+ */
+
+AuthorizeHandler.prototype.getState = function(request) {
+  var state = request.body.state || request.query.state;
+
+  if (!state) {
+    throw new InvalidRequestError('Missing parameter: `state`');
+  }
+
+  if (!is.vschar(state)) {
+    throw new InvalidRequestError('Invalid parameter: `state`');
+  }
+
+  return state;
+};
+
+/**
+ * Get user by calling the authenticate middleware.
+ */
+
+AuthorizeHandler.prototype.getUser = function(request, response) {
+  return this.authenticateHandler.handle(request, response).then(function(token) {
```

Comment:
The [OAuth spec](https://tools.ietf.org/html/rfc6749#page-25) leaves the decision of granting the access request up to the server application:

> The authorization server … establishes whether the resource owner grants or denies the client's access request.

This is functionally equivalent to the previous implementation found [here](https://github.com/thomseddon/node-oauth2-server/blob/master/lib/authCodeGrant.js#L138) with the exception that no extra callback is given to the server app. Instead, an assumption is made that the user-agent must be previously authenticated in order to grant an authorization code.

This has worked quite well for us and we found that, overall, it reduces the burden of the implementation of the authorization code flow by using the same authentication mechanism that the server app has presumably already implemented (if they're using the `node-oauth-server` module, that is).